cpp/daal/src/algorithms/service_kernel_math.h

```diff
@@ -673,72 +684,152 @@ bool solveEquationsSystemWithCholesky(FPType * a, FPType * b, size_t n, size_t n
 }
 
 template <typename FPType, CpuType cpu>
-bool solveEquationsSystemWithPLU(FPType * a, FPType * b, size_t n, size_t nX, bool sequential, bool extendFromSymmetric)
+bool solveEquationsSystemWithSpectralDecomposition(FPType * a, FPType * b, size_t n, size_t nX, bool sequential)
 {
-    if (extendFromSymmetric)
+    /* Storage for the eigenvalues.
+    Note: this allocates more size than they might require when nX > 1, because the same
+    buffer will get reused later on and needs the extra size. Those additional entries
+    will not be filled with eigenvalues. */
+    TArrayScalable<FPType, cpu> eigenvalues(n * nX);
+    DAAL_CHECK_MALLOC(eigenvalues.get());
+
+    /* SYEV parameters */
+    char jobz = 'V';
+    char uplo = 'U';
+    DAAL_INT info;
+
+    /* Query the procedure for size of required buffer */
+    DAAL_INT lwork = -1;
+    FPType buffer_size;
+    if (sequential)
     {
-        /* Extend symmetric matrix to generic through filling of upper triangle */
-        for (size_t i = 0; i < n; ++i)
-        {
-            for (size_t j = 0; j < i; ++j)
-            {
-                a[j * n + i] = a[i * n + j];
-            }
-        }
+        LapackInst<FPType, cpu>::xxsyev(&jobz, &uplo, (DAAL_INT *)&n, a, (DAAL_INT *)&n, eigenvalues.get(), &buffer_size, &lwork, &info);
     }
 
-    /* GETRF and GETRS parameters */
-    char trans    = 'N';
-    DAAL_INT info = 0;
+    else
+    {
+        LapackInst<FPType, cpu>::xsyev(&jobz, &uplo, (DAAL_INT *)&n, a, (DAAL_INT *)&n, eigenvalues.get(), &buffer_size, &lwork, &info);
+    }
 
-    TArrayScalable<DAAL_INT, cpu> ipiv(n);
-    DAAL_CHECK_MALLOC(ipiv.get());
+    if (info) return false;
 
-    /* Perform P*L*U factorization of A */
+    /* Check that buffer size will not overflow when passed to LAPACK */
+    if (static_cast<size_t>(buffer_size) > std::numeric_limits<DAAL_INT>::max()) return false;
+
+    /* Allocate work buffer as needed */
+    DAAL_INT work_buffer_size = static_cast<DAAL_INT>(buffer_size);
+    TArrayScalable<FPType, cpu> work_buffer(work_buffer_size);
+    DAAL_CHECK_MALLOC(work_buffer.get());
+
+    /* Perform Q*diag(l)*Q' factorization of A */
     if (sequential)
     {
-        LapackInst<FPType, cpu>::xxgetrf((DAAL_INT *)&n, (DAAL_INT *)&n, a, (DAAL_INT *)&n, ipiv.get(), &info);
+        LapackInst<FPType, cpu>::xxsyev(&jobz, &uplo, (DAAL_INT *)&n, a, (DAAL_INT *)&n, eigenvalues.get(), work_buffer.get(), &work_buffer_size,
```

Comment:
Thanks for the suggestion. I'm not sure if I'm interpreting the docs correctly. There's this phrase:
> Note that for most cases of real symmetric eigenvalue problems the default choice should be [syevr](https://www.intel.com/content/www/us/en/docs/onemkl/developer-reference-c/2024-2/syevr.html#GUID-8A87F3BA-0F1C-488A-B0A5-E0BFD7D99841) function

.. which is a bit ambiguous as it can be interpreted in two ways:
* Users should prefer calling `syevr` instead of `syev`.
* `syev` will end up calling `syevr` in most cases.

I see that [SciPy defaults to syevr](https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.eigh.html#scipy.linalg.eigh) so perhaps the first interpretation is the correct one, would be ideal to have some confirmation.

Reply:
Found an interesting paper right at Netlib:
https://www.netlib.org/lapack/lawnspdf/lawn183.pdf

It compares the symmetric eigenvalue routines in Netlib's LAPACK so not sure how generalizable it is to MKL, but it appears that there's a tradeoff between speed and accuracy of these routines (`syev`, `syevd`, `syevr`, `syevx`).

From what they describe, `syevr` looks to be consistently ~50% faster than `syev`, but has much larger errors (by orders of magniture depending on the data size), whereas `syevd` which is used elsewhere in `oneDAL` has the smallest errors.

Since the procedure here is discarding small eigenvalues in the first place, I guess these numerical inaccuracies should not be much of a concern, and thus `syevr` should be preferred as it's the fastest. Will modify the PR to call it instead.

Reply:
Changed to `syevr`.